pkg/kubelet/status/manager.go

```diff
@@ -323,8 +323,14 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
 	// TODO: make me easier to express from client code
 	pod, err := m.kubeClient.Pods(status.podNamespace).Get(status.podName)
 	if errors.IsNotFound(err) {
-		glog.V(3).Infof("Pod %q (%s) was deleted on the server", status.podName, uid)
-		m.deletePodStatus(uid)
+		glog.V(3).Infof("Pod %q (%s) does not exist on the server", status.podName, uid)
+		// Note that we do not delete the pod status because the pod to update may
+		// be a static pod whose mirror pod is in the process of recreation.
+		// Deleting the status without clearing the corresponding
+		// apiStatusVersions entry may lead to wrong versioning and stale status.
```

Comment:
You got me there. The problem extends beyond that, and applies to mirror/static pods only.

Let's walk through one example. The local version of static pod A is 4, with both version 3 and 4 queued up in the channel. The apiStatusVersion is 2.

1. Mirror pod gets deleted.
2. SyncBatch checks the cache and the apiStatusVersion, decides to send version 4 to the apiserver. It could not find the mirror pod, and decides to reset the local version to 0.
3. Mirror pod gets created.
4. syncPod starts to process update version 3 from the channel, and successfully sends the update to the apiserver. apiStatusVersion becomes 3.

Now the local version is 0 and the apiStatusVersion is 3.